File: insight/src/main/java/info/nightscout/androidaps/plugins/pump/insight/connection_service/MessageRequest.java
package info.nightscout.androidaps.plugins.pump.insight.connection_service;

import info.nightscout.androidaps.plugins.pump.insight.app_layer.AppLayerMessage;

public class MessageRequest<T extends AppLayerMessage> implements Comparable<MessageRequest> {

    T request;
    T response;
    Exception exception;

    MessageRequest(T request) {
        this.request = request;
    }

    public T await() throws Exception {
        synchronized (this) {
            while (exception == null && response == null) wait();
            if (exception != null) throw exception;
            return response;
        }
    }

    public T await(long timeout) throws Exception {
        synchronized (this) {
            while (exception == null && response == null) wait(timeout);
            if (exception != null) throw exception;
            return response;
        }
    }

    @Override
    public int compareTo(MessageRequest messageRequest) {
        return request.compareTo(messageRequest.request);
    }

    public T getRequest() {
        return this.request;
    }

    public T getResponse() {
        return this.response;
    }

    public Exception getException() {
        return this.exception;
    }
}

File: pump/insight/src/main/kotlin/app/aaps/pump/insight/connection_service/MessageRequest.kt
package app.aaps.pump.insight.connection_service

import app.aaps.core.utils.wait
import app.aaps.core.utils.waitMillis
import app.aaps.pump.insight.app_layer.AppLayerMessage

open class MessageRequest<T : AppLayerMessage> internal constructor(var request: T) : Comparable<MessageRequest<*>> {

    var response: T? = null
    var exception: Exception? = null

    @Throws(Exception::class)
    open fun await(): T {
        synchronized(this) {
            while (exception == null && response == null) wait()
            exception?.let { e -> throw e }
            return response!!
        }
    }

    @Throws(Exception::class)
    open fun await(timeout: Long): T {
        synchronized(this) {
            while (exception == null && response == null) waitMillis(timeout)
            exception?.let { e -> throw e }
            return response!!
        }
    }

    override fun compareTo(other: MessageRequest<*>): Int {
        return request.compareTo(other.request)
    }
}